newhelm/cache_helper.py

```diff
@@ -0,0 +1,87 @@
+# mypy: ignore-errors
+from abc import ABC, abstractmethod
+import json
+import os
+from pydantic import BaseModel
+from sqlitedict import SqliteDict  # type: ignore
+from typing import Any, Dict, Generic, Mapping, TypeVar, Optional, get_args
+
+from newhelm.sut import SUT
+from newhelm.typed_data import TypedData
+
+RequestType = TypeVar("RequestType", bound=BaseModel)
+ResponseType = TypeVar("ResponseType", bound=BaseModel)
+
+
+class CacheHelper(ABC, Generic[RequestType, ResponseType]):
+    @abstractmethod
+    def get_cached_response(self, request: RequestType) -> Optional[ResponseType]:
+        """Returns the SUT's cached response to a request."""
+
+    @abstractmethod
+    def update_cache(self, request: RequestType, response: ResponseType):
+        """Cache the SUT's response in-memory."""
+
+
+class SUTResponseCacheHelper(CacheHelper[RequestType, ResponseType]):
```

Comment:
My understanding is generics require you to know the types at code writing time, and we can't do that for caching.

If we want strong typing from `get_cached_response` I think we could do something like:

```
get_cached_response(request, sut: BasePromptResponseSUT[Request,Response]) -> Response`
```

We could also just drop the typing fully and simplify all of this.

Reply:
If we dropped the typing, how would we know what ResponseType the responses should be converted to?

Reply:
In my drive to make TypedData do strong typing, I totally missed that it could use reflection to reconstruct itself without the caller needing to know the type. #142 makes that work.